action/protocol/staking/protocol.go

```diff
@@ -114,3 +147,55 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name CandName) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) validateDelegate(d *Delegate) error {
+	if _, err := address.FromString(d.Owner); err != nil {
+		return errors.Wrap(err, "invalid owner address")
+	}
+
+	if _, err := address.FromString(d.Address); err != nil {
+		return errors.Wrap(err, "invalid operator address")
+	}
+
+	if _, err := address.FromString(d.RewardAddress); err != nil {
+		return errors.Wrap(err, "invalid reward address")
+	}
+
+	if d.Votes.Sign() <= 0 {
+		return ErrInvalidAmount
+	}
+	return nil
+}
+
+func (p *Protocol) addDelegate(d *Delegate) error {
+	if err := p.validateDelegate(d); err != nil {
+		return err
+	}
+
+	if !p.hasDelegate(d.CanName) {
+		p.delegates[d.CanName] = d
+	}
+	return nil
+}
+
+func (p *Protocol) delDelegate(name CandName) {
+	delete(p.delegates, name)
+}
+
+func (p *Protocol) addVote(name CandName, amount *big.Int) error {
+	if p.hasDelegate(name) {
+		return p.delegates[name].AddVote(amount)
+	}
+	return nil
```

Comment:
there will be a problem:
0. we have delegate [d1, d2, d3, d4]
1. voter votes delegate "d_not_exist"
2. someone registered as "d_not_exist"

Reply:
this need to be changed per latest discussion

first do for the in-memory map, then try entry in DB